papermill/engines.py

```diff
@@ -0,0 +1,328 @@
+import sys
+import copy
+import datetime
+import dateutil
+
+from functools import wraps
+
+# tqdm creates 2 globals lock which raise OSException if the execution
+# environment does not have shared memory for processes, e.g. AWS Lambda
+try:
+    from tqdm import tqdm, tqdm_notebook
+
+    no_tqdm = False
+except OSError:
+    no_tqdm = True
+
+from .exceptions import PapermillException
+from .preprocess import PapermillExecutePreprocessor
+from .iorw import write_ipynb
+
+
+class PapermillEngines(object):
+    """
+    The holder which houses any engine registered with the system.
+    This object is used in a singleton manner to save and load particular
+    named EngineBase objects for reference externally.
+    """
+
+    def __init__(self):
+        self._engines = {}
+
+    def register(self, name, engine):
+        self._engines[name] = engine
+
+    def get_engine(self, name=None):
+        """
+        Retrieves an engine by name.
+        """
+        engine = self._engines.get(name)
+        if not engine:
+            raise PapermillException("No engine named '{}' found".format(name))
+        return engine
+
+    def execute_notebook_with_engine(self, engine_name, nb, kernel_name, **kwargs):
+        """
+        Fetches the approiate engine and executes the nb object against it.
+        """
+        return self.get_engine(engine_name).wrap_and_execute_notebook(nb, kernel_name, **kwargs)
+
+
+def catch_nb_assignment(func):
+    """
+    Wrapper which helps catch `nb` keyword arguments and assign onto self
+    when passed to the wrapped function.
+    Used for callback methods when the caller may optionally have a new copy
+    of the originally wrapped `nb`object.
+    """
+
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+        nb = kwargs.get('nb')
+        if nb:
+            # Reassign if executing notebook object was replaced
+            self.nb = nb
+        return func(self, *args, **kwargs)
+
+    return wrapper
+
+
+class EngineNotebookWrapper(object):
```

Comment:
It took me several attempts to get what this piece does and how it is used. Now I think of it as thing that hold metadata and state about the execution progress of a notebook. It provides some hooks I can call to update the state as I (the engine) churns through the notebook.

I don't know if a different name would have helped me get this idea sooner or not. Some suggestions: `NotebookExecutionState` or `EngineProgressState`? Wrapper makes me think we wrap some layer around the notebook to make it compatible with how an engine expects it to look.

Reply:
I was struggling with this class name. Here's a few more ideas `EngineExecutionTracker`, `EngineProgressor`, `EngineExecutionState`? I might be leaning towards your suggestion of `EngineProgressState`. Anyone else have suggestions / think one of these sounds good.

Reply:
One more `EngineState`? (I actually like `EngineProgressState` the least, progress and state are too similar, feels like `StripedZebra` :-/)

Reply:
I don't think progress and state are too similar — progress suggests tracking where it is in it's trajectory, state merely indicates what the current state is.

Reply:
Remember that it's through this class that the engine can access the notebook. I imagine that's at least as important to mention in naming the class as the progress tracking part?

It's a notebook viewed through progress-tracking-tinted glasses. I can't think of a name that doesn't sound silly... A `ProgressingNotebook`?